File: LinkedList/MaxTwinSum/maxTwin.cpp
#include <iostream>
#include <climits>
using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverse(ListNode* head){
        ListNode* prev = NULL;
        ListNode* curr = head;
        ListNode* Next = head;
        while(curr){
            Next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = Next;
        }
        return prev;
    }
    int pairSum(ListNode* head) {
        int maxSum = INT_MIN;
        ListNode* temp = head;
        int n = 0;
        while(temp){
            temp = temp->next;
            n++;
        }
        temp = head;
        ListNode* help = head;
        int check = n;
        while(help){
            if(n==check/2)break;
            help = help->next;
            n--;
        }
        help = reverse(help);
        while(help){
            maxSum = max(help->val+temp->val,maxSum);
            help = help->next;
            temp = temp->next;
        }
        return maxSum;
    }
};

File: LinkedList/MaxTwinSum/maxTwin.py
from typing import Optional
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def reverseList(self, head):
        # Code here
        prev = None
        curr = head
        next = head
        while curr:
            next = curr.next
            curr.next = prev
            prev = curr
            curr = next
        return prev
    def pairSum(self, head: Optional[ListNode]) -> int:
        slow = head
        fast = head
        while fast.next and fast.next.next:
            slow = slow.next
            fast = fast.next.next
        revPart = self.reverseList(slow.next)
        slow.next = None
        temp = head
        maxSum = -float('inf')
        while temp:
            currSum = temp.val + revPart.val
            maxSum = max(maxSum,currSum)
            revPart = revPart.next
            temp = temp.next
        return maxSum

